### PR TITLE
[EM-1497] Add Google Pay support

### PR DESCRIPTION
**Pull request by kan98**
Add support for creating a token with Google Pay

https://user-images.githubusercontent.com/34004612/156327942-25e1f430-d204-4053-b20d-c5f1aff70db0.mp4

### CODE REVIEW

File: .idea/compiler.xml [hunk 1]
```diff
@@ -1,22 +1,6 @@
...
-      <entry name="!?*.aj" />
-    </wildcardResourcePatterns>
-    <annotationProcessing>
-      <profile default="true" name="Default" enabled="false">
-        <processorPath useClasspath="true" />
-      </profile>
-    </annotationProcessing>
+    <bytecodeTargetLevel target="1.8" />
   </component>
 </project>
```
[the remark this answers is not in the document]
Comment (kan98) [reply .idea/compiler.xml:6]: this file is automatically generated by Android studio (intellij), I think we can leave it as it is.

**Comment by nuxzero** [comment]
@kan98 Is there any breaking changes?

**Comment by kan98** [comment]
> @kan98 Is there any breaking changes?

@nuxzero no kub, there should only be an addition of the new flow. Everything else should work the way it does in the current version.

File: sdk/src/test/java/co/omise/android/TokenizationMethodTest.kt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    //
+
+    @Test
+    fun googlepay_isReadyToPayRequest_returnsCorrectValues() {
+        val googlepay = Googlepay("pkey_123", arrayListOf("Visa"), 3000, "thb", "merchantId")
+        val isReadyToPay = googlepay.isReadyToPayRequest()
+
+
+        val expected = JSONObject(
+                "{\"apiVersion\":2,\"apiVersionMinor\":0,\"allowedPaymentMethods\":[{\"type\":\"CARD\",\"parameters\":{\"allowedAuthMethods\":[\"PAN_ONLY\"],\"allowedCardNetworks\":[\"VISA\"]}}]}"
```
Comment (nuxzero) [inline sdk/src/test/java/co/omise/android/TokenizationMethodTest.kt:71]: In Kotlin, You can use multiple lines string like below.

```kotlin
val json = """
{
	"apiVersion": 2
}
"""
"""
```

File: app/src/main/java/co/omise/android/example/PaymentSetting.kt [hunk 3]
```diff
@@ -100,6 +102,17 @@ object PaymentSetting {
...
+        val tokenizationMethods = paymentMethodPreferences
+                .filter { it.value }
+                .toMap()
+                .map {
+                    when (it.key) {
+                        context.getString(R.string.payment_preference_googlepay_key) -> TokenizationMethod.GooglePay
+                        else -> null
+                    }
+                }
+                .filterNotNull()
```
Comment (jutinapas) [inline app/src/main/java/co/omise/android/example/PaymentSetting.kt:114]: Not sure if it works but maybe `mapNotNull` can be used for combining `.map` and `.filterNotNull`.

File: sdk/src/test/java/co/omise/android/TokenizationMethodTest.kt (hunk 2 above)
Comment (kan98) [reply sdk/src/test/java/co/omise/android/TokenizationMethodTest.kt]: wow I'm starting to like Kotlin now 😅 
changed here kub https://github.com/omise/omise-android/pull/213/commits/2166f006af8e9754e2e8ab3fb88d03e18bfdf908

File: sdk/src/main/java/co/omise/android/models/Token.kt [hunk 4]
```diff
@@ -9,6 +9,7 @@ import okhttp3.HttpUrl
 import okhttp3.RequestBody
 import org.joda.time.DateTime
 import java.io.IOException
+import kotlin.jvm.Throws
```
[the remark this answers is not in the document]
Comment (kan98) [reply sdk/src/main/java/co/omise/android/models/Token.kt:12]: line 46 raises a warning if this isn't imported kub

File: sdk/build.gradle [hunk 5]
```diff
@@ -162,6 +162,7 @@ dependencies {
     implementation "androidx.lifecycle:lifecycle-livedata-ktx:$lifecycle_version"
     implementation "org.jetbrains.kotlinx:kotlinx-coroutines-core:$coroutines_version"
     implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:$coroutines_version"
+    implementation 'com.google.android.gms:play-services-wallet:19.1.0'
```
[the remark this answers is not in the document]
Comment (kan98) [reply sdk/build.gradle:165]: it's required for us to accept google pay payments. What would be the best thing to do in this case kub?

File: sdk/src/main/java/co/omise/android/models/GooglePay.kt [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+     */
+    private fun allowedCardNetworks(): JSONArray {
+        val networksMapping = HashMap<String, String>()
+        networksMapping["American Express"] = "AMEX"
+        networksMapping["JCB"] = "JCB"
+        networksMapping["MasterCard"] = "MASTERCARD"
+        networksMapping["Visa"] = "VISA"
+
+        var newList = arrayListOf<String>()
+        if (!this.cardNetworks.isEmpty())
```
Comment (hpariwat) [inline sdk/src/main/java/co/omise/android/models/GooglePay.kt:66]: How about `.isNotEmpty()`

File: sdk/src/main/java/co/omise/android/models/GooglePay.kt [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+) {
+    private val gateway = "omise"
+
+    /**
+     * Create a Google Pay API base request object with properties used in all requests.
+     *
+     * @return Google Pay API base request object.
+     * @throws JSONException
+     */
+    private val baseRequest = JSONObject().apply {
```
[the remark this answers is not in the document]
Comment (kan98) [reply sdk/src/main/java/co/omise/android/models/GooglePay.kt:27]: boiler plate code written by Google used the standard library implementation so I just stuck with it. Is it worth switching over for maintainability purposes?

File: sdk/src/test/java/co/omise/android/TokenizationMethodTest.kt [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+        assertEquals(TokenizationMethod.Unknown("aaa"), tokenizationMethod)
+    }
+
+    //
+    // Google Pay Tests
+    //
+
+    @Test
+    fun googlepay_isReadyToPayRequest_returnsCorrectValues() {
```
[the remark this answers is not in the document]
Comment (kan98) [reply sdk/src/test/java/co/omise/android/TokenizationMethodTest.kt]: moved to a new file here https://github.com/omise/omise-android/pull/213/commits/64484971d2a6033ebb1644115d1c0cdefc60c8c1

File: sdk/src/main/java/co/omise/android/models/GooglePay.kt [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+    /**
+     * Card networks supported by your app and your gateway.
+     *
+     *
+     * @return Allowed card networks
+     * @see [CardParameters](https://developers.google.com/pay/api/android/reference/object.CardParameters)
+     */
+    private fun allowedCardNetworks(): JSONArray {
+        val networksMapping = HashMap<String, String>()
```
[the remark this answers is not in the document]
Comment (kan98) [reply sdk/src/main/java/co/omise/android/models/GooglePay.kt]: changed here https://github.com/omise/omise-android/pull/213/commits/64484971d2a6033ebb1644115d1c0cdefc60c8c1

File: sdk/src/main/java/co/omise/android/models/Capability.kt [hunk 10]
```diff
@@ -45,18 +59,20 @@ data class Capability(
...
          * @param allowCreditCard allow to create a [Token] with a credit card or not. Default is true.
          * @param sourceTypes list of [SourceType] that allow to create a [Source].
+         * @param tokenizationMethods list of [TokenizationMethod] that we can create a [Token] with.
          * @param zeroInterestInstallments whether merchant absorbs interest for installment payments.
          *
          * @return an instance of [Capability] with specific configuration.
          */
         @JvmStatic
-        fun create(allowCreditCard: Boolean = true, sourceTypes: List<SourceType>, zeroInterestInstallments: Boolean = false): Capability {
+        fun create(allowCreditCard: Boolean = true, sourceTypes: List<SourceType>, tokenizationMethods: List<TokenizationMethod>, zeroInterestInstallments: Boolean = false): Capability {
```
[the remark this answers is not in the document]
Comment (kan98) [reply sdk/src/main/java/co/omise/android/models/Capability.kt]: added here https://github.com/omise/omise-android/pull/213/commits/64484971d2a6033ebb1644115d1c0cdefc60c8c1

File: .idea/compiler.xml [hunk 11]
```diff
@@ -1,22 +1,6 @@
...
-      <entry name="!?*.kt" />
-      <entry name="!?*.clj" />
-      <entry name="!?*.aj" />
-    </wildcardResourcePatterns>
-    <annotationProcessing>
-      <profile default="true" name="Default" enabled="false">
-        <processorPath useClasspath="true" />
-      </profile>
-    </annotationProcessing>
+    <bytecodeTargetLevel target="1.8" />
```
[the remark this answers is not in the document]
Comment (kan98) [reply .idea/compiler.xml]: undid the changes here https://github.com/omise/omise-android/pull/213/commits/64484971d2a6033ebb1644115d1c0cdefc60c8c1

File: sdk/src/main/java/co/omise/android/models/GooglePay.kt (hunk 6 above)
Comment (kan98) [reply sdk/src/main/java/co/omise/android/models/GooglePay.kt]: changed here https://github.com/omise/omise-android/pull/213/commits/64484971d2a6033ebb1644115d1c0cdefc60c8c1

**Review by hpariwat — APPROVED** [review]

**Review by new4762 — APPROVED** [review]

File: sdk/src/main/java/co/omise/android/ui/GooglePayActivity.kt [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+import com.google.android.gms.common.api.ApiException
+import com.google.android.gms.wallet.*
+import com.google.android.material.snackbar.Snackbar
+import kotlinx.android.synthetic.main.activity_credit_card.*
+import kotlinx.android.synthetic.main.activity_google_pay.*
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.IOError
+
+class GooglePayActivity : AppCompatActivity() {
```
[the remark this answers is not in the document]
Comment (nuxzero) [reply sdk/src/main/java/co/omise/android/ui/GooglePayActivity.kt:26]: That's okay not big deal. Please update the docs to register `GooglePayactivity` to their apps.

File: sdk/src/test/java/co/omise/android/GooglePayTest.kt [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+import co.omise.android.request.GooglePay
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.json.JSONObject
+import org.junit.Assert.assertEquals
+import org.junit.Test
+import org.junit.runner.RunWith
+
+
+@RunWith(AndroidJUnit4::class)
+class GooglePayTest {
```
Comment (nuxzero) [inline sdk/src/test/java/co/omise/android/GooglePayTest.kt:13]: Usually, class and unit test class should have same package name. Could you make a change?

**Review by nuxzero — APPROVED** [review]
The code looks good to me. However, there is just a small comment in test package name. I gave an approval anyway.

File: sdk/src/test/java/co/omise/android/GooglePayTest.kt (hunk 13 above)
Comment (kan98) [reply sdk/src/test/java/co/omise/android/GooglePayTest.kt]: changed here https://github.com/omise/omise-android/pull/213/commits/9b6b980e09601648c2a7b31592d736498f4cc471

**Review by tookaa2 — APPROVED** [review]

**Pull request merged**